zeus/monitor/carbon.py

```diff
@@ -394,39 +397,36 @@ def _update_carbon_emissions(key: str):
         now = datetime.now(timezone.utc)
         hour_floor = now.replace(minute=0, second=0, microsecond=0)
         hour_ceil = hour_floor + timedelta(hours=1)
-        hour_key = hour_floor.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
 
         # start windows
         for key in keys:
-            zeus_monitor.begin_window(key)
+            zeus_monitor.begin_window(key, sync_execution=False)
 
         try:
             # continuously fetch from command q until hit hour
             while now < hour_ceil:
                 seconds_until_hour = (hour_ceil - now).total_seconds()
                 op, key = command_q.get(timeout=seconds_until_hour)
 
-                if op == CarbonEmissionMonitor.Op.BEGIN:
-                    zeus_monitor.begin_window(key)
+                if op == Op.BEGIN:
+                    zeus_monitor.begin_window(key, sync_execution=False)
                     keys.add(key)
-                elif op == CarbonEmissionMonitor.Op.END:
+                elif op == Op.END:
                     # update if has not been updated in a while
-                    _update_energy_measurements(key, hour_key)
+                    _update_energy_measurements(key, hour_floor)
                     _update_carbon_emissions(key)
                     finished_q.put(
                         (
-                            key,
-                            dict(gpu_carbon_emissions[key]),
-                            dict(cpu_carbon_emissions[key]),
+                            dict(gpu_carbon_emissions.pop(key)),
+                            dict(cpu_carbon_emissions.pop(key)),
+                            dict(dram_carbon_emissions.pop(key)),
                         )
                     )
                     keys.remove(key)
-                    del gpu_carbon_emissions[key]
-                    del cpu_carbon_emissions[key]
 
                     if len(keys) == 0:
                         return
-                elif op == CarbonEmissionMonitor.Op.NEXTITER:
+                elif op == Op.NEXTITER:
```

Comment:
Just wanted to mention that this is a very neat idea!

Reply:
Thank you!